### PR TITLE
Integrate the whole pie

### PR DESCRIPTION
**Pull request by quite**

### CODE REVIEW

**Review by secworks — APPROVED** [review]

**Review by secworks — APPROVED** [review]

**Review by secworks — APPROVED** [review]

**Review by secworks — APPROVED** [review]

**Review by secworks — APPROVED** [review]

File: hw/application_fpga/fw/tk1/main.c [hunk 1]
```diff
@@ -115,18 +115,20 @@ static void compute_cdi(const uint8_t digest[32], const uint8_t use_uss,
 {
 	uint32_t local_cdi[8];
 	blake2s_ctx secure_ctx;
+	uint32_t rnd;
```
Comment (secworks) [inline hw/application_fpga/fw/tk1/main.c:97]: Nits: Use a better name than just rnd. rnd_sleep for example.

**Review by secworks — APPROVED** [review]

**Review by secworks — APPROVED** [review]

File: hw/application_fpga/fw/tk1/main.c [hunk 2]
```diff
@@ -244,7 +244,11 @@ int main()
...
 		// Now we know the size of the cmd frame, read it all
-		read(cmd, hdr.len);
+		if (read(cmd, CMDLEN_MAXBYTES, hdr.len) != 0) {
+			htif_puts("read: buffer overrun\n");
+			forever_redflash();
+			// Not reached
+		}
 
 		// Is it for us?
 		if (hdr.endpoint != DST_FW) {
```
Comment (secworks) [inline hw/application_fpga/fw/tk1/main.c]: Do we still have endpoint info in the messages?

File: hw/application_fpga/fw/tk1/main.c [hunk 3]
```diff
@@ -108,30 +108,26 @@ static void compute_cdi(uint8_t digest[32], uint8_t use_uss, uint8_t uss[32])
...
 	}
 
 	// Write hashed result to Compound Device Identity (CDI)
-	blake2s_final(secure_ctx, local_cdi);
+	blake2s_final(&secure_ctx, local_cdi);
 
-	// Write over the firmware-only RAM
-	memset((void *)fw_ram, 0, TK1_MMIO_FW_RAM_SIZE);
+	// Clear secure_ctx of any residue
+	memset(&secure_ctx, 0, sizeof(secure_ctx));
```
Comment (secworks) [inline hw/application_fpga/fw/tk1/main.c:129]: Isn't the whole fw_ram cleared before jumping to the app? What is left?

**Review by secworks — APPROVED** [review]

File: hw/application_fpga/fw/tk1/main.c (hunk 1 above)
Comment (mchack-work) [reply hw/application_fpga/fw/tk1/main.c:97]: Ok.

File: hw/application_fpga/fw/tk1/main.c (hunk 2 above)
Comment (mchack-work) [reply hw/application_fpga/fw/tk1/main.c]: Yes, and we decided to use these endpoint things in the signer TKey program as well when a client program like tkey-ssh-agent probes for firmware or signer.

File: hw/application_fpga/fw/tk1/main.c (hunk 3 above)
Comment (mchack-work) [reply hw/application_fpga/fw/tk1/main.c:129]: Yes, entire `fw_ram` is cleared by inline assembler before jumping. This is just being careful.

**Pull request merged**